### ISSUE TITLE
Error when Using the package in python process

### ISSUE THREAD

**Issue by anibo12** [rstudio/reticulate#972]
Sorry if anyone else has raised this issue before. I am working on a package migration project from R to Python and I need to have the python object to be accessible inside the embedded R from the Python process. However, reticulate package keeps getting segfault,

> Python 3.6.10 |Anaconda, Inc.| (default, Jan  7 2020, 21:14:29) 
> [GCC 7.3.0] on linux
> Type "help", "copyright", "credits" or "license" for more information.

> >import rpy2.robjects as robjects
> >import rpy2.robjects.packages as rpackages
> >r_env = rpackages.importr("reticulate")
> 
> >robjects.r['import']("pandas")
> R[write to console]: 
>  *** caught segfault ***
> 
> R[write to console]: address 0x10, cause 'memory not mapped'
> 
> R[write to console]: 
> Traceback:
> 
> R[write to console]:  1: 
> R[write to console]: py_initialize(config$python, config$libpython, config$pythonhome, 
> R[write to console]:     config$virtualenv_activate, config$version >= "3.0", interactive(), 
> R[write to console]:     numpy_load_error)
> R[write to console]: 
> 
> R[write to console]:  2: 
> R[write to console]: doTryCatch(return(expr), name, parentenv, handler)
> R[write to console]: 
> 
> R[write to console]:  3: 
> R[write to console]: tryCatchOne(expr, names, parentenv, handlers[[1L]])
> R[write to console]: 
> 
> R[write to console]:  4: 
> R[write to console]: tryCatchList(expr, classes, parentenv, handlers)
> R[write to console]: 
> 
> R[write to console]:  5: 
> R[write to console]: tryCatch({
> R[write to console]:     py_initialize(config$python, config$libpython, config$pythonhome, 
> R[write to console]:         config$virtualenv_activate, config$version >= "3.0", 
> R[write to console]:         interactive(), numpy_load_error)
> R[write to console]: }, error = function(e) {
> R[write to console]:     Sys.setenv(PATH = oldpath)
> R[write to console]:     if (is.na(curr_session_env)) {
> R[write to console]:         Sys.unsetenv("R_SESSION_INITIALIZED")
> R[write to console]:     }
> R[write to console]:     else {
> R[write to console]:         Sys.setenv(R_SESSION_INITIALIZED = curr_session_env)
> R[write to console]:     }
> R[write to console]:     stop(e)
> R[write to console]: })
> R[write to console]: 
> 
> R[write to console]:  6: 
> R[write to console]: initialize_python(required_module, use_environment)
> R[write to console]: 
> 
> R[write to console]:  7: 
> R[write to console]: ensure_python_initialized(required_module = module)
> R[write to console]: 
> 
> R[write to console]:  8: 
> R[write to console]: (function (module, as = NULL, convert = TRUE, delay_load = FALSE) 
> R[write to console]: {
> R[write to console]:     if (!is.null(as)) {
> R[write to console]:         register_class_filter(function(classes) {
> R[write to console]:             sub(paste0("^", module), as, classes)
> R[write to console]:         })
> R[write to console]:     }
> R[write to console]:     delay_load_environment <- NULL
> R[write to console]:     delay_load_priority <- 0
> R[write to console]:     delay_load_functions <- NULL
> R[write to console]:     if (is.function(delay_load)) {
> R[write to console]:         delay_load_functions <- list(on_load = delay_load)
> R[write to console]:         delay_load <- TRUE
> R[write to console]:     }
> R[write to console]:     else if (is.list(delay_load)) {
> R[write to console]:         delay_load_environment <- delay_load$environment
> R[write to console]:         delay_load_functions <- delay_load
> R[write to console]:         if (!is.null(delay_load$priority)) 
> R[write to console]:             delay_load_priority <- delay_load$priority
> R[write to console]:         delay_load <- TRUE
> R[write to console]:     }
> R[write to console]:     if (!delay_load || is_python_initialized()) {
> R[write to console]:         ensure_python_initialized(required_module = module)
> R[write to console]:         hookName <- paste("reticulate", module, "load", sep = "::")
> R[write to console]:         imported <- py_module_import(module, convert = convert)
> R[write to console]:         hooks <- getHook(hookName)
> R[write to console]:         for (hook in hooks) tryCatch(hook(imported), error = warning)
> R[write to console]:         setHook(hookName, list(), "replace")
> R[write to console]:         imported
> R[write to console]:     }
> R[write to console]:     else {
> R[write to console]:         if (is.null(.globals$delay_load_module) || (delay_load_priority > 
> R[write to console]:             .globals$delay_load_priority)) {
> R[write to console]:             .globals$delay_load_module <- module
> R[write to console]:             .globals$delay_load_environment <- delay_load_environment
> R[write to console]:             .globals$delay_load_priority <- delay_load_priority
> R[write to console]:         }
> R[write to console]:         module_proxy <- new.env(parent = emptyenv())
> R[write to console]:         module_proxy$module <- module
> R[write to console]:         module_proxy$convert <- convert
> R[write to console]:         if (!is.null(delay_load_functions)) {
> R[write to console]:             module_proxy$get_module <- delay_load_functions$get_module
> R[write to console]:             module_proxy$on_load <- delay_load_functions$on_load
> R[write to console]:             module_proxy$on_error <- delay_load_functions$on_error
> R[write to console]:         }
> R[write to console]:         attr(module_proxy, "class") <- c("python.builtin.module", 
> R[write to console]:             "python.builtin.object")
> R[write to console]:         module_proxy
> R[write to console]:     }
> R[write to console]: })("pandas")
> R[write to console]: 
> 
> R[write to console]: 
> Possible actions:
> 1: abort (with core dump, if enabled)
> 2: normal R exit
> 3: exit R without saving workspace
> 4: exit R saving workspace
> 
> Selection: 1
> R[write to console]: R is aborting now ...
> 
> Segmentation fault (core dumped)

Is there a way to avoid this kind of errors?

**Comment by toddcreasy** [comment]
Hi @anibo12, Did anyone ever look into this issue? I'm experiencing the same problem trying to run a scanpy pipeline.

rpy2:
```
 rcode = '''
        rmarkdown::render("scanpy.Rmd")
 '''
robjects.r(rcode)

```
rmarkdown where the error occurs:

```
{python import, echo=TRUE, include=TRUE, warning=TRUE, results='asis'}
import scanpy as sc
adata = sc.read("scanpyz.h5ad")
summary = adata.uns["summary"]
summary

```

segfault

  |............................                                          |  40%
label: import (with options)
List of 5
 $ echo   : logi TRUE
 $ include: logi TRUE
 $ warning: logi TRUE
 $ results: chr "asis"
 $ engine : chr "python"

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:
 *** caught segfault ***

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: address 0x10, cause 'memory not mapped'

**Comment by kevinushey** [comment]
We will need a reproducible example to be able to learn more.

**Comment by toddcreasy** [comment]
@kevinushey 

You will need to download pbmc3k_processed.h5ad (or any other h5ad anndata file).

You can get that [here at 10X](url). You might need to fill out a short form. 

Let me know if this is not enough for you to reproduce.

# python code

```
import rpy2.robjects as robjects
import rpy2.robjects.packages as rpackages
from rpy2.robjects.packages import importr
from rpy2.robjects.vectors import StrVector`

def create_seurat_object():
    """ Create Seurat object using rpy2 """
    # import R "base" package
    base = importr("base")

    # import R "utils" package
    utils = importr("utils")

    # select a mirror for R packages
    utils.chooseCRANmirror(ind=1)  # select the first mirror in the list

    # R package names
    packnames = ('reticulate', 'Seurat', 'Matrix')

    # Selectively install what needs to be install.
    # We are fancy, just because we can.
    names_to_install = [x for x in packnames if not rpackages.isinstalled(x)]
    if len(names_to_install) > 0:
        utils.install_packages(StrVector(names_to_install))

    rcode = '''
            rmarkdown::render("scanpy.Rmd")
            '''
    robjects.r(rcode)```


```
Rmarkdown code (scanpy.Rmd)

\-\-
output: 
  html_document
\-\-\-

\`\`\`{r setup, echo=TRUE, include=FALSE, warning = FALSE, results='asis'}

suppressPackageStartupMessages({
    library("reticulate")
    library("Seurat")
    library("knitr")
})

options(max.print="75")
knitr::opts_chunk$set(echo=FALSE,
                      prompt=FALSE,
                      dev="png",
                      dev.args = list(type = "cairo-png"),
                      tidy=TRUE,
                      comment=NA,
                      message=FALSE,
                      warning=FALSE)

opts_knit$set(width=1200)

\`\`\`

\`\`\`{python import, echo=TRUE, include=TRUE, warning=TRUE, results='asis'}
import scanpy as sc
adata = sc.read("pbmc3k_processed.h5ad")
summary = adata.uns["summary"]
summary
\`\`\`
```

**Comment by toddcreasy** [comment]
Here is the output I get:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

processing file: scanpy.Rmd

  |..............                                                        |  20%
  ordinary text without R code

  |............................                                          |  40%
label: setup (with options)
List of 4
 $ echo   : logi TRUE
 $ include: logi FALSE
 $ warning: logi FALSE
 $ results: chr "asis"

  |..........................................                            |  60%
  ordinary text without R code

  |........................................................              |  80%
label: import (with options)
List of 5
 $ echo   : logi TRUE
 $ include: logi TRUE
 $ warning: logi TRUE
 $ results: chr "asis"
 $ engine : chr "python"

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:
 *** caught segfault ***

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: address 0x10, cause 'memory not mapped'

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:
Traceback:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  1:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: py_initialize(config$python, config$libpython, config$pythonhome,
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     config$virtualenv_activate, config$version >= "3.0", interactive(),
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     numpy_load_error)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  2:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: (function() {
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     Sys.setenv(PYTHONPATH = newpythonpath)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     on.exit(Sys.setenv(PYTHONPATH = oldpythonpath), add = TRUE)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     py_initialize(config$python, config$libpython, config$pythonhome,
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         config$virtualenv_activate, config$version >= "3.0",
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         interactive(), numpy_load_error)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: })()
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  3:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: doTryCatch(return(expr), name, parentenv, handler)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  4:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: tryCatchOne(expr, names, parentenv, handlers[[1L]])
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  5:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: tryCatchList(expr, classes, parentenv, handlers)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  6:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: tryCatch({
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     oldpythonpath <- Sys.getenv("PYTHONPATH")
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     newpythonpath <- Sys.getenv("RETICULATE_PYTHONPATH", unset = paste(config$pythonpath,
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         system.file("python", package = "reticulate"), sep = .Platform$path.sep))
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     local({
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         Sys.setenv(PYTHONPATH = newpythonpath)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         on.exit(Sys.setenv(PYTHONPATH = oldpythonpath), add = TRUE)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         py_initialize(config$python, config$libpython, config$pythonhome,
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:             config$virtualenv_activate, config$version >= "3.0",
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:             interactive(), numpy_load_error)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     })
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: }, error = function(e) {
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     Sys.setenv(PATH = oldpath)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     if (is.na(curr_session_env)) {
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         Sys.unsetenv("R_SESSION_INITIALIZED")
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     }
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     else {
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         Sys.setenv(R_SESSION_INITIALIZED = curr_session_env)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     }
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     stop(e)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: })
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  7:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: initialize_python(required_module, use_environment)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  8:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: ensure_python_initialized()
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  9:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: eng_python_initialize(options = options, envir = environment())
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 10:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: reticulate::eng_python(options)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 11:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: engine(options)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 12:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: in_dir(input_dir(), engine(options))
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 13:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: block_exec(params)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 14:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: call_block(x)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 15:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: process_group.block(group)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 16:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: process_group(group)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 17:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: withCallingHandlers(if (tangle) process_tangle(group) else process_group(group),
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     error = function(e) {
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         setwd(wd)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         cat(res, sep = "\n", file = output %n% "")
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:         message("Quitting from lines ", paste(current_lines(i),
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:             collapse = "-"), " (", knit_concord$get("infile"),
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:             ") ")
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     })
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 18:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: process_file(text, output)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 19:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: knitr::knit(knit_input, knit_output, envir = envir, quiet = quiet)
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 20:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: rmarkdown::render("scanpy.Rmd")
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 21:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: (function (expr, envir = parent.frame(), enclos = if (is.list(envir) ||
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     is.pairlist(envir)) parent.frame() else baseenv())
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: .Internal(eval(expr, envir, enclos)))(expression(rmarkdown::render("scanpy.Rmd")))
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 22:
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: (function (expr, envir = parent.frame(), enclos = if (is.list(envir) ||
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:     is.pairlist(envir)) parent.frame() else baseenv())
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: .Internal(eval(expr, envir, enclos)))(expression(rmarkdown::render("scanpy.Rmd")))
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:
Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace

**Comment by t-kalinowski** [comment]
Closed via #1188
